### PR TITLE
[PM-6789] rc version legacy migration fix

### PR DESCRIPTION
**Pull request by jlf0dev**
Copied from #9498, we unfortunately had key rotation changes that were merged immediately after `rc` was cut. This is a blocker for our release, so I have created a branch off the original work and rebased it on `rc`, without the new key rotation changes.

Everything should be the exact same as #9498, before commit 20b5edca6dfe7a20729b9f782b52805f02777d9a. Which means everything was previously pre-approved (even though I may not be able to prove it 😞).

## 🎟️ Tracking

https://bitwarden.atlassian.net/browse/PM-6789
## 📔 Objective

Fixes legacy user migration by using the legacy key for private key re-encryption.

I took the chance to finish distributing key rotation logic to each domain using an interface.
## 📸 Screenshots

https://github.com/bitwarden/clients/assets/24985544/48c0d98f-4625-4de5-8f6e-7f634f2568ea

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

File: libs/common/src/tools/send/services/send.service.spec.ts [hunk 1]
```diff
@@ -400,36 +400,31 @@ describe("SendService", () => {
...
-      // eslint-disable-next-line @typescript-eslint/no-floating-promises
-      sendService.replace(null);
-
-      await awaitAsync();
-
-      const newUserKey = new SymmetricCryptoKey(new Uint8Array(32)) as UserKey;
-      const result = await sendService.getRotatedKeys(newUserKey);
-
-      expect(result).toEqual([]);
+    it("throws if the original user key is null", async () => {
```
Comment (audrey-jensen) [inline libs/common/src/tools/send/services/send.service.spec.ts:420]: ⚠️ The return value of `getRotatedData` is no longer tested. Please restore that test.

**Review by audrey-jensen — CHANGES_REQUESTED** [review]
Sorry, just noticed a problem on the RC PR that also applies to this one.

File: libs/common/src/tools/send/services/send.service.spec.ts (hunk 1 above)
Comment (audrey-jensen) [inline libs/common/src/tools/send/services/send.service.spec.ts:420]: ⚠️ The return value of `getRotatedData` is no longer tested. Please restore that test.

**Review by audrey-jensen — CHANGES_REQUESTED** [review]

File: libs/common/src/tools/send/services/send.service.spec.ts (hunk 1 above)
Comment (jlf0dev) [reply libs/common/src/tools/send/services/send.service.spec.ts:433]: Please see b36e5ca225d6e00db20ae4a49f32ce38b970c282

Comment (audrey-jensen) [reply libs/common/src/tools/send/services/send.service.spec.ts:433]: Perfect!

**Review by audrey-jensen — APPROVED** [review]

**Review by JaredSnider-Bitwarden — APPROVED** [review]

**Review by addisonbeck — APPROVED** [review]

**Review by coroiu — APPROVED** [review]

**Review by Jingo88 — APPROVED** [review]

**Pull request merged**